llpc/context/llpcCompiler.cpp

```diff
@@ -728,6 +728,246 @@ template <typename T> static FormattedBytes formatBytesLittleEndian(ArrayRef<T>
   return format_bytes(bytes, /* FirstByteOffset = */ None, /* NumPerLine = */ 16, /* ByteGroupSize = */ 1);
 }
 
+// Build shader from the specified info.
+//
+// @param [in]  pipelineInfo : Info to build this shader module
+// @param [out] pipelineOut : Output of building this shader module
+// @param [in]  stage : Shader stage of needing to compile
+// @param pPipelineDumpFile : Handle of pipeline dump file
+//
+// @returns : Result::Success if successful. Other return codes indicate failure.
+Result Compiler::buildGraphicsShaderStage(const GraphicsPipelineBuildInfo *pipelineInfo,
+                                          GraphicsPipelineBuildOut *pipelineOut, UnlinkedShaderStage stage,
+                                          void *pPipelineDumpFile) {
+  Result result = Result::Success;
+
+  BinaryData elfBin = {};
+  // clang-format off
+  SmallVector<const PipelineShaderInfo *, ShaderStageGfxCount> shaderInfo = {
+    &pipelineInfo->vs,
+    &pipelineInfo->tcs,
+    &pipelineInfo->tes,
+    &pipelineInfo->gs,
+    &pipelineInfo->fs,
+  };
+  // clang-format on
+
+  MetroHash::Hash cacheHash = {};
+  MetroHash::Hash pipelineHash = {};
+  cacheHash = PipelineDumper::generateHashForGraphicsPipeline(pipelineInfo, true, true);
+  pipelineHash = PipelineDumper::generateHashForGraphicsPipeline(pipelineInfo, true, true);
+
+  if (pPipelineDumpFile) {
+    std::stringstream strStream;
+    strStream << ";Compiler Options: ";
+    for (auto &option : m_options)
+      strStream << option << " ";
```

Comment:
If you use `raw_string_ostream` you can join all options with `llvm::join`.

Reply:
what are the differences between them?

Reply:
There's not that much difference here. `raw_string_ostream` is much simpler than stringstream (in terms of the implementation and the API) and has a bunch of utility functions made with llvm types in mind. What I'm suggesting would be something like:

```c++
std::string buf;
raw_string_ostream os;
os << ";Compiler Options: " << join(m_options, " ");
os.flush();
PipelineDumper::DumpPipelineExtraInfo(reinterpret_cast<PipelineDumpFile *>(pipelineDumpFile), &buf);
```

Reply:
Also, this could be a helper (member) function, if it's used elsewhere in this file.

Reply:
Add a member function.